`cx-freeze` fails to install since #7271

This fails to install for me:
```
  - Updating cx-freeze (7.0.0 876fe77 -> 7.1.1 876fe77): Failed

  ChefBuildError

  Backend subprocess exited when trying to invoke build_editable
  
  running editable_wheel
  creating /tmp/poetry-chef-l92dv9z4/.tmp-zywo1jiy/cx_Freeze.egg-info
  writing /tmp/poetry-chef-l92dv9z4/.tmp-zywo1jiy/cx_Freeze.egg-info/PKG-INFO
  writing dependency_links to /tmp/poetry-chef-l92dv9z4/.tmp-zywo1jiy/cx_Freeze.egg-info/dependency_links.txt
  writing entry points to /tmp/poetry-chef-l92dv9z4/.tmp-zywo1jiy/cx_Freeze.egg-info/entry_points.txt
  writing requirements to /tmp/poetry-chef-l92dv9z4/.tmp-zywo1jiy/cx_Freeze.egg-info/requires.txt
  writing top-level names to /tmp/poetry-chef-l92dv9z4/.tmp-zywo1jiy/cx_Freeze.egg-info/top_level.txt
  writing manifest file '/tmp/poetry-chef-l92dv9z4/.tmp-zywo1jiy/cx_Freeze.egg-info/SOURCES.txt'
  reading manifest file '/tmp/poetry-chef-l92dv9z4/.tmp-zywo1jiy/cx_Freeze.egg-info/SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  adding license file 'LICENSE.md'
  writing manifest file '/tmp/poetry-chef-l92dv9z4/.tmp-zywo1jiy/cx_Freeze.egg-info/SOURCES.txt'
  creating '/tmp/poetry-chef-l92dv9z4/.tmp-zywo1jiy/cx_Freeze-7.1.1.dist-info'
  creating /tmp/poetry-chef-l92dv9z4/.tmp-zywo1jiy/cx_Freeze-7.1.1.dist-info/WHEEL
  running build_py
  running build_ext
  creating /tmp/tmp28l8tueg.build-temp/source
  creating /tmp/tmp28l8tueg.build-temp/source/bases
  x86_64-linux-gnu-gcc -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -fPIC -I/tmp/tmp5avh93ft/.venv/include -I/usr/include/python3.10 -c source/bases/console.c -o /tmp/tmp28l8tueg.build-temp/source/bases/console.o
  source/bases/console.c:7:10: fatal error: Python.h: No such file or directory
      7 | #include <Python.h>
        |          ^~~~~~~~~~
  compilation terminated.
  Traceback (most recent call last):
    File "/tmp/tmp5avh93ft/.venv/lib/python3.10/site-packages/setuptools/_distutils/spawn.py", line 71, in spawn
      subprocess.check_call(cmd, env=_inject_macos_ver(env))
    File "/usr/lib/python3.10/subprocess.py", line 369, in check_call
      raise CalledProcessError(retcode, cmd)
  subprocess.CalledProcessError: Command '['/usr/bin/x86_64-linux-gnu-gcc', '-Wno-unused-result', '-Wsign-compare', '-DNDEBUG', '-g', '-fwrapv', '-O2', '-Wall', '-g', '-fstack-protector-strong', '-Wformat', '-Werror=format-security', '-g', '-fwrapv', '-O2', '-fPIC', '-I/tmp/tmp5avh93ft/.venv/include', '-I/usr/include/python3.10', '-c', 'source/bases/console.c', '-o', '/tmp/tmp28l8tueg.build-temp/source/bases/console.o']' returned non-zero exit status 1.
  
  The above exception was the direct cause of the following exception:
  
  Traceback (most recent call last):
    File "/tmp/tmp5avh93ft/.venv/lib/python3.10/site-packages/setuptools/_distutils/unixccompiler.py", line 191, in _compile
      self.spawn(compiler_so + cc_args + [src, '-o', obj] + extra_postargs)
    File "/tmp/tmp5avh93ft/.venv/lib/python3.10/site-packages/setuptools/_distutils/ccompiler.py", line 1044, in spawn
      spawn(cmd, dry_run=self.dry_run, **kwargs)
    File "/tmp/tmp5avh93ft/.venv/lib/python3.10/site-packages/setuptools/_distutils/spawn.py", line 77, in spawn
      raise DistutilsExecError(
  distutils.errors.DistutilsExecError: command '/usr/bin/x86_64-linux-gnu-gcc' failed with exit code 1
  
  During handling of the above exception, another exception occurred:
  
  Traceback (most recent call last):
    File "/tmp/tmp5avh93ft/.venv/lib/python3.10/site-packages/setuptools/command/editable_wheel.py", line 153, in run
      self._create_wheel_file(bdist_wheel)
    File "/tmp/tmp5avh93ft/.venv/lib/python3.10/site-packages/setuptools/command/editable_wheel.py", line 355, in _create_wheel_file
      files, mapping = self._run_build_commands(dist_name, unpacked, lib, tmp)
    File "/tmp/tmp5avh93ft/.venv/lib/python3.10/site-packages/setuptools/command/editable_wheel.py", line 278, in _run_build_commands
      self._run_build_subcommands()
    File "/tmp/tmp5avh93ft/.venv/lib/python3.10/site-packages/setuptools/command/editable_wheel.py", line 305, in _run_build_subcommands
      self.run_command(name)
    File "/tmp/tmp5avh93ft/.venv/lib/python3.10/site-packages/setuptools/_distutils/cmd.py", line 316, in run_command
      self.distribution.run_command(command)
    File "/tmp/tmp5avh93ft/.venv/lib/python3.10/site-packages/setuptools/dist.py", line 974, in run_command
      super().run_command(command)
    File "/tmp/tmp5avh93ft/.venv/lib/python3.10/site-packages/setuptools/_distutils/dist.py", line 989, in run_command
      cmd_obj.run()
    File "<string>", line 261, in run
    File "/tmp/tmp5avh93ft/.venv/lib/python3.10/site-packages/setuptools/command/build_ext.py", line 93, in run
      _build_ext.run(self)
    File "/tmp/tmp5avh93ft/.venv/lib/python3.10/site-packages/setuptools/_distutils/command/build_ext.py", line 359, in run
      self.build_extensions()
    File "/tmp/tmp5avh93ft/.venv/lib/python3.10/site-packages/setuptools/_distutils/command/build_ext.py", line 479, in build_extensions
      self._build_extensions_serial()
    File "/tmp/tmp5avh93ft/.venv/lib/python3.10/site-packages/setuptools/_distutils/command/build_ext.py", line 505, in _build_extensions_serial
      self.build_extension(ext)
    File "<string>", line 46, in build_extension
    File "/tmp/tmp5avh93ft/.venv/lib/python3.10/site-packages/setuptools/_distutils/ccompiler.py", line 605, in compile
      self._compile(obj, src, ext, cc_args, extra_postargs, pp_opts)
    File "/tmp/tmp5avh93ft/.venv/lib/python3.10/site-packages/setuptools/_distutils/unixccompiler.py", line 193, in _compile
      raise CompileError(msg)
  distutils.errors.CompileError: command '/usr/bin/x86_64-linux-gnu-gcc' failed with exit code 1
  /tmp/tmp5avh93ft/.venv/lib/python3.10/site-packages/setuptools/_distutils/dist.py:989: _DebuggingTips: Problem in editable installation.
  !!
  
          ********************************************************************************
          An error happened while installing `cx_Freeze` in editable mode.
  
          The following steps are recommended to help debug this problem:
  
          - Try to install the project normally, without using the editable mode.
            Does the error still persist?
            (If it does, try fixing the problem before attempting the editable mode).
          - If you are using binary extensions, make sure you have all OS-level
            dependencies installed (e.g. compilers, toolchains, binary libraries, ...).
          - Try the latest version of setuptools (maybe the error was already fixed).
          - If you (or your project dependencies) are using any setuptools extension
            or customization, make sure they support the editable mode.
  
          After following the steps above, if the problem still persists and
          you think this is related to how setuptools handles editable installations,
          please submit a reproducible example
          (see https://stackoverflow.com/help/minimal-reproducible-example) to:
  
              https://github.com/pypa/setuptools/issues
  
          See https://setuptools.pypa.io/en/latest/userguide/development_mode.html for details.
          ********************************************************************************
  
  !!
    cmd_obj.run()
  error: command '/usr/bin/x86_64-linux-gnu-gcc' failed with exit code 1
```

GPT-4o says it's because you need the `python3-dev` package for this. We shouldn't require that from users.

IMO this build stuff should be in an `extra` dependency set rather than the main dependencies.

_Originally posted by @Pwuts in https://github.com/Significant-Gravitas/AutoGPT/issues/7271#issuecomment-2201302412_

--- a/autogpt/pyproject.toml
+++ b/autogpt/pyproject.toml
@@ -43,10 +43,15 @@ sentry-sdk = "^1.40.4"
 # Benchmarking
 agbenchmark = { path = "../benchmark", optional = true }
 # agbenchmark = {git = "https://github.com/Significant-Gravitas/AutoGPT.git", subdirectory = "benchmark", optional = true}
-cx-freeze = { git = "https://github.com/ntindle/cx_Freeze.git", rev = "main", develop = true }
+
+# Build
+cx-freeze = { git = "https://github.com/ntindle/cx_Freeze.git", rev = "main", optional = true }
+# HACK: switch to cx-freeze release package after #2442 and #2472 are merged: https://github.com/marcelotduarte/cx_Freeze/pulls?q=is:pr+%232442+OR+%232472+
+# cx-freeze = { version = "^7.2.0", optional = true }
 
 [tool.poetry.extras]
 benchmark = ["agbenchmark"]
+build = ["cx-freeze"]
 
 [tool.poetry.group.dev.dependencies]
 black = "^23.12.1"

--- a/autogpt/pyproject.toml
+++ b/autogpt/pyproject.toml
@@ -68,8 +68,11 @@ pytest-recording = "*"
 pytest-xdist = "*"
 vcrpy = { git = "https://github.com/Significant-Gravitas/vcrpy.git", rev = "master" }
 
+[tool.poetry.group.build]
+optional = true
+
 [tool.poetry.group.build.dependencies]
-cx-freeze = { git = "https://github.com/ntindle/cx_Freeze.git", rev = "main", optional = true }
+cx-freeze = { git = "https://github.com/ntindle/cx_Freeze.git", rev = "main" }
 # HACK: switch to cx-freeze release package after #2442 and #2472 are merged: https://github.com/marcelotduarte/cx_Freeze/pulls?q=is:pr+%232442+OR+%232472+
 # cx-freeze = { version = "^7.2.0", optional = true }
 

--- a/autogpt/pyproject.toml
+++ b/autogpt/pyproject.toml
@@ -44,14 +44,8 @@ sentry-sdk = "^1.40.4"
 agbenchmark = { path = "../benchmark", optional = true }
 # agbenchmark = {git = "https://github.com/Significant-Gravitas/AutoGPT.git", subdirectory = "benchmark", optional = true}
 
-# Build
-cx-freeze = { git = "https://github.com/ntindle/cx_Freeze.git", rev = "main", optional = true }
-# HACK: switch to cx-freeze release package after #2442 and #2472 are merged: https://github.com/marcelotduarte/cx_Freeze/pulls?q=is:pr+%232442+OR+%232472+
-# cx-freeze = { version = "^7.2.0", optional = true }
-
 [tool.poetry.extras]
 benchmark = ["agbenchmark"]
-build = ["cx-freeze"]
 
 [tool.poetry.group.dev.dependencies]
 black = "^23.12.1"
@@ -73,6 +67,11 @@ pytest-mock = "*"
 pytest-recording = "*"
 pytest-xdist = "*"
 vcrpy = { git = "https://github.com/Significant-Gravitas/vcrpy.git", rev = "master" }
+
+[tool.poetry.group.build.dependencies]
+cx-freeze = { git = "https://github.com/ntindle/cx_Freeze.git", rev = "main", optional = true }
+# HACK: switch to cx-freeze release package after #2442 and #2472 are merged: https://github.com/marcelotduarte/cx_Freeze/pulls?q=is:pr+%232442+OR+%232472+
+# cx-freeze = { version = "^7.2.0", optional = true }
 
 
 [build-system]

